### DISCUSSION TITLE
reverse record

### DISCUSSION THREAD

**Discussion by lukatendai** [coredns/coredns#5672] [General]
I have been using tinydns for ages and one thing that I like is that  the reverse record is created automatically.  Can I do this the same in coredns?

**Comment by lukatendai** [comment]
Also, I am experimenting and I wonder why hosts does not generate revers record in this case:

```
test.local:1053 {
        bind lo enp0s3
        records {
                @   60  IN MX 10 mx1
                mx1 60  IN A  127.0.0.1
        }
        hosts {
                127.0.0.2 www.test.local
                fallthrough
        }
}
```

This is response:

```
➜  ~ nslookup -port=1053 mx1.test.local 192.168.86.194
Server:		192.168.86.194
Address:	192.168.86.194#1053

Name:	mx1.test.local
Address: 127.0.0.1

➜  ~ nslookup -port=1053 www.test.local 192.168.86.194
Server:		192.168.86.194
Address:	192.168.86.194#1053

Name:	www.test.local
Address: 127.0.0.2

➜  ~ nslookup -port=1053 127.0.0.2 192.168.86.194    
** server can't find 2.0.0.127.in-addr.arpa: NXDOMAIN


```

**Comment by chrisohaver** [comment]
CoreDNS capability is too dynamic to make this possible globally. But for certain plugins, where all the records are known it is possible. The hosts plugin for instance does this. Although hosts has to be configured handle the reverse zone too (in your example it only handles test.local.).

**Reply by lukatendai** [reply]
Thank you for your reply. Can you please explain "Although hosts has to be configured handle the reverse zone too (in your example it only handles test.local.)."  What should be the configuration of test.local so 127.0.0.2 resolves to www.test.local?

**Comment by chrisohaver** [comment]
The way reverse lookups work in DNS is the client sends a query for the IP address in the reverse zone (for ipv4 the reverse zone is `in-addr.arpa.`).  An example reverse lookup query for 127.0.0.2 would be `2.0.0.127.in-addr.arpa.` (the ip octet order is reversed).  For CoreDNS to be able to answer queries in the `in-addr.arpa` domain, it needs to know which server block should handle it. This has to be configured manually. for example:

```
test.local in-addr.arpa {
        hosts {
                127.0.0.2 www.test.local
        }
}
```

**Reply by lukatendai** [reply]
I understated. And, yest it works in that case. however when I define new zone, e.g.

```
test.local:1053 in-addr.arpa {
        bind lo enp0s3
        records {
                @   60  IN MX 10 mx1
                mx1 60  IN A  127.0.0.1
        }
        hosts {
                127.0.0.2 www.test.local
                fallthrough
        }
}

test2.local:1053 in-addr.arpa {
        bind lo enp0s3
        records {
                @   60  IN MX 10 mx1
                mx1 60  IN A  127.0.0.1
        }
        hosts {
                127.0.0.2 www.test2.local
                fallthrough
        }
}
```

then I get error that  in-addr.arpa is already defined

We are managing about 100 domains and I wish I could do similar thing what I do in tinydns, when I keep all records about specific domain in one place, e.g.

```
.test.local:192.168.1.2:ns-01.test.local:600
.test.local:192.168.1.3:ns-02.test.local:600
@test.local:192.168.1.2:a::600
=test.local:192.168.1.2:600
=t.test.local:192.168.1.3:600
+www.test.local:192.168.1.2:600
+mail.test.local:192.168.1.2:600

.test2.local:192.168.1.4:ns-01.test2.local:600
.test2.local:192.168.1.5:ns-02.test2.local:600
@test2.local:192.168.1.4:a::600
=test2.local:192.168.1.4:600
=t.test2.local:192.168.1.5:600
+www.test2.local:192.168.1.4:600
+mail.test2.local:192.168.1.4:600
```

How I can achieve it in coredns?

**Comment by chrisohaver** [comment]
I think the example is a bit contrived, since both ip addresses are identical.

Nevertheless, I think this does what you you want:

```
test.local:1053 test2.local:1053 in-addr.arpa:1053 {
        bind lo enp0s3
        hosts {
                127.0.0.2 www.test.local www.test2.local
                fallthrough
        }
}
```

**Reply by lukatendai** [reply]
Thank you. Yeah, I thought that I can keep each domain in a separate file so it is easy to manage...

**Comment by chrisohaver** [comment]
> I thought that I can keep each domain in a separate file

If the ips are in different subnets, then you can.

```
test.local:1053 1.10.in-addr.arpa:1053 {
        bind lo enp0s3
        hosts {
                10.1.0.1 www.test.local
        }
}

test2.local:1053 2.10.in-addr.arpa:1053 {
        bind lo enp0s3
        hosts {
                10.2.0.1 www.test2.local
        }
}
```
